File: java/parsers/src/main/java/jp/co/teruuu/parser/json/JArrayParser.java
package jp.co.teruuu.parser.json;

import jp.co.teruuu.parser.common.Pair2;
import jp.co.teruuu.parser.common.Pair3;
import jp.co.teruuu.parser.common.ParseResult;
import jp.co.teruuu.parser.common.Parser;
import jp.co.teruuu.parser.common.type.Tuple2;
import jp.co.teruuu.parser.common.type.Tuple3;
import jp.co.teruuu.parser.json.type.JArray;
import jp.co.teruuu.parser.json.type.JString;
import jp.co.teruuu.parser.json.type.JValue;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;

public class JArrayParser implements JParser<JValue> {
    Parser<List<String>> sp = Parser.string(" ").many();
    Parser<Tuple3<List<String>, String, List<String>>> left = new Pair3(sp, Parser.string("["), sp);
    Parser<Tuple3<List<String>, String, List<String>>> right = new Pair3(sp, Parser.string("]"), sp);

    public JArrayParser() {
    }

    @Override
    public ParseResult<JValue> parse(String input) {
        String next = input;
        ParseResult leftResult = this.left.parse(next);
        if(leftResult instanceof ParseResult.Success) {
            next = ((ParseResult.Success) leftResult).next;
            Parser<Optional<Tuple2<Tuple2<JValue, List<String>>, List<Tuple2<String, Tuple3<List<String>, JValue, List<String>>>>>>> parser =
                    (new Pair2(new Pair2(JParser.value(), sp), (new Pair2(Parser.string(","), new Pair3(sp, JParser.value(), sp))).many())).option();
            ParseResult result = parser.parse(next);
            if(result instanceof ParseResult.Success) {
                next = ((ParseResult.Success) result).next;
                ParseResult rightResult = this.right.parse(next);

                if(rightResult instanceof ParseResult.Success) {
                    Optional<Tuple2<Tuple2<JValue, List<String>>, List<Tuple2<String, Tuple3<List<String>, JValue, List<String>>>>>> resultValue =
                            ((ParseResult.Success<Optional<Tuple2<Tuple2<JValue, List<String>>, List<Tuple2<String, Tuple3<List<String>, JValue, List<String>>>>>>>) result).value;
                    next = ((ParseResult.Success) rightResult).next;
                    JArray ret = new JArray();
                    if(resultValue.isPresent()) {
                        ret.add(resultValue.get().item1.item1);
                        for(Tuple2<String, Tuple3<List<String>, JValue, List<String>>> a: resultValue.get().item2) {
                            ret.add(a.item2.item2);
                        }
                    }
                    return new ParseResult.Success(ret, next);
                } else {
                    return rightResult;
                }
            } else {
                return result;
            }
        } else {
            return leftResult;
        }
    }
}

File: scala/parsers/src/main/scala/jp/co/teruuu/parser/json/JArrayParser.scala
package jp.co.teruuu.parser.json

import jp.co.teruuu.parser.common._
import jp.co.teruuu.parser.json.`type`.{JArray, JValue}

object JArrayParser extends JParser[JArray] {
  lazy val sp = Parser.string(" ").many()
  lazy val leftP = new Pair3(sp, Parser.string("["), sp)
  lazy val rightP = new Pair3(sp, Parser.string("]"), sp)
  lazy val parser = (new Pair2(new Pair3(sp, JValueParser, sp), (new Pair2(Parser.string(","), (new Pair3(sp, JValueParser, sp)))).many())).option()


  override def parse(input: String): ParseResult[JArray] = {
    leftP.parse(input) match {
      case leftResult: ParseOk[(List[String], String, List[String])] => {
        parser.parse(leftResult.next) match {
          case result: ParseOk[Option[((List[String], JValue, List[String]), List[(String, (List[String], JValue, List[String]))])]] => {
            rightP.parse(result.next) match {
              case rightResult: ParseOk[(List[String], String, List[String])] => {
                var value = List.empty[JValue]
                if(result.value.isDefined) {
                  value = value :+ result.value.get._1._2
                  result.value.get._2.foreach(a => {
                    value = value :+ a._2._2
                  })
                }
                ParseOk(JArray(value), rightResult.next)
              }
              case rightResult: ParseNg[_] => ParseNg(rightResult.message, rightResult.next)
            }
          }
          case result: ParseNg[_] => ParseNg(result.message, result.next)
        }
      }
      case leftResult: ParseNg[_] => ParseNg(leftResult.message, leftResult.next)
    }
  }
}